### ISSUE TITLE
Speech not working on Dynamic Edit on initial opening

### ISSUE THREAD

**Issue by jmwmusic** [musescore/MuseScore#33573]
### Issue type

Accessibility issue (e.g. for keyboard-only or screen reader users)

### Description with steps to reproduce

When trying to edit an existing dynamic mark the screen reader gives no speech until something is deleted/changed.

STEPS:
1.  Add a dynamic mark such as mp (best to show with more than 2 characters)
2. Use Alt + arrows to navigate on to the dynamic mark and press enter.
3. Press left and right arrows and notice there is no speech output.
4. Arrow left a few times to ensure that you are at the beginning of the dynamic mark and press delete twice to remove it.
5. Type a new dynamic mark such as fp and notice that the screen reader announces as these are entered.
6. Arrow left and right across the newly entered dynamic and notice that they now do read each character correctly as you move across them.  Until you reach the start or the end where instead of stating blank or saying nothing they continue to repeat the previous letter (See #19933 which is particularly confusing if you are trying to tell whether you have ff or fff.

The screen reader should announce each character as you arrow along when the field first opens.

### Supporting files, videos and screenshots

N/A

### In which versions of MuseScore Studio is this issue present?

4.7

### Regression

Yes, this used to work in a previous version of MuseScore 4.x

### Operating system

Windows 11 with NVDA

### Checklist

- [x] This report follows the [guidelines](https://github.com/musescore/MuseScore/wiki/Reporting-bugs-and-issues) for reporting bugs and issues
- [x] I have verified that this issue has not been logged before, by searching the [issue tracker](https://github.com/musescore/MuseScore/issues) for similar issues
- [x] I have attached all requested files and information to this report
- [x] I have attempted to identify the root problem as concisely as possible, and have used minimal reproducible examples where possible

**Comment by shoogle** [comment]
Dynamic characters have never been accessible in MuseScore so this isn't a regression. Rather, it's a symptom of a wider issue where SMuFL characters are not spoken by screen readers.

When you type `fp` in Dynamic entry mode, you're typing ASCII text characters, which the screen reader knows how to read.

As soon as leave Dynamic edit mode (e.g. by pressing Esc), the `f` and `p` are converted to SMuFL characters. SMuFL isn't part of the Unicode standard, so screen readers don't know what to do when they encounter these characters. As far as the screen reader is concerned, they are just arbitrary codepoints within Unicode's [private use area](https://en.wikipedia.org/wiki/Private_Use_Areas), which could have any meaning.

(Automatic conversion of ASCII text to SMuFL dynamics is fairly new, which might explain your confusion. But it's always been possible to enter SMuFL dynamics via the Palette, so the bug with screen reader speech is not new.)

For dynamics, the fix is straightforward: we just need to replace SMuFL characters with their ASCII counterparts in the accessible text we send to the screen reader.

* SMuFL **_p_** &rarr; ASCII `p`
* SMuFL **_f_** &rarr; ASCII `f`
* Etc.

For other SMuFL characters it's less clear. A symbol like [quarter-tone-sharp](https://w3c-cg.github.io/smufl/latest/tables/other-accidentals.html) has no ASCII or unicode equivalent, and even if it did, the screen reader probably wouldn't know how to pronounce it.

We might have to replace the SMuFL character with literal text, like this:

* SMuFL **quarter-tone-sharp** (a single character) &rarr; `quarter-tone-sharp` (multiple characters)

Which would be a bit like:

* `+` (symbol) &rarr; `plus sign` (literal text)

The problem is the screen reader needs to be aware of the cursor position and the selection state. As you press the right arrow key to move through the text, at some point the cursor will move over the SMuFL character (or over the `+`). When that happens, we would have to tell the screen reader that you moved past the literal string that replaced the SMuFL character (or `plus sign`). In other words, the screen reader will think the cursor jumped over multiple characters instead of just one.

**Comment by jmwmusic** [comment]
Thanks for explaining what is happening here.  Would this solution also work to fix #19984?  Even if it is a more limited set of the most common note values in tempo marks.

**Comment by shoogle** [comment]
Yes, it would likely work for that. We could send something like this to the screen reader:

* Shorthand: `q = 150`
* Explicit: `quarter-note = 50`

We could provide an option in Preferences to control this, or to disable such replacements entirely and rely on the screen reader to make sense of the raw SMuFL codepoint.
